src/ECL/ECL.hs

```diff
@@ -68,23 +105,36 @@ genericMerge f bsl@((s, b):bs) bsr@((s', b'):bs')
 
 -- | Returns the uncovered set after one binder is applied to the set of
 -- values represented by another.
-missingSingle :: (Eq lit) => Binder lit -> Binder lit -> Binders lit
-missingSingle _ (Var _) = []
-missingSingle b@(Var _) (Tagged tag bc) =
-  map (Tagged tag) $ missingSingle b bc
-missingSingle c@(Tagged tag b) (Tagged tag' b')
-  | tag == tag' = map (Tagged tag) $ missingSingle b b'
-  | otherwise = [c]
-missingSingle (Var _) (Record bs) =
-  map (Record . zip names) miss
+missingSingle :: (Eq lit) => Environment -> Binder lit -> Binder lit -> ([Binder lit], Maybe Bool)
+missingSingle _ _ (Var _) = ([], pure True)
+missingSingle env (Var _) (Product ps) =
+  first (map Product) $ missingMultiple env (initialize $ length ps) ps
+missingSingle env p@(Product ps) (Product ps')
+  | length ps == length ps' = first (map Product) $ missingMultiple env ps ps'
+  | otherwise = ([p], pure False)
+missingSingle env (Var _) cb@(Tagged con _) =
+  (concatMap (\cp -> fst $ missingSingle env cp cb) $ tagEnv, pure True)
   where
-  miss = getUncovered $ missingMultiple (initialize $ length bs) binders
+  tag :: (Eq lit) => (Name, Arity) -> Binder lit
+  tag (n, 0) = Tagged n $ wildcard
```

Comment:
Well it would, but I'm saying `Tagged s (Var Nothing)` is a perfectly fine representation for something which has tag `s`, but for which we know nothing else.